### PR TITLE
[feat]Qwen3.6-35B-A3B 支持--speculative-algorithm DSPARK

### PR DESCRIPTION
**Pull request by chenyuege**
## Motivation

本 MR 修复 `--speculative-algorithm DSPARK` 在 Qwen3.6-35B-A3B 上使用不同 draft checkpoint 时的兼容性问题，使以下两类权重能够在 HCU 环境下完成加载、推理和精度评测：

- `Qwen3.6-35B-A3B-speculator.dspark`[https://huggingface.co/RedHatAI/Qwen3.6-35B-A3B-speculator.dspark]：使用缩减词表、独立 draft LM head 和 `d2t` 映射。

修复依据 checkpoint 配置、权重形状和内核能力进行判断，不依赖模型目录名。部署保留主模型及 draft 的 `--attention-backend fa3`、`--kv-cache-dtype fp8_e5m2`、CUDA Graph，以及 `SGLANG_ROCM_USE_AITER_MOE=1`。

基线为 `upstream/release/20260825_v0.5.18`（`6d59319229`），工作分支为 `feat/Qwen3.6-35B-A3B-dpark-0.5.18`。本 MR 包含以下三个依次提交的源码修复，最终 HEAD 为 `48fcb0299b`。

### 1. `2c6b570779` — 修复 reduced-vocabulary draft 加载与 token 映射

Commit：`2c6b5707792d5d25bfde722af5297ef645c25ca5`

`fix(dspark): load reduced vocabulary heads and map target token logits`

**问题：** 旧 draft 的词表为 32000，目标模型词表为 248320。Checkpoint 的 Markov 输入权重为 `[248320,256]`，而原实现建立的参数为 `[32000,256]`，导致加载失败。此外，缩减词表的 token ID 不能直接作为目标模型 token ID，独立 draft LM head 也不能被共享的 target LM head 替代。

**修改：**

- 在 `python/sglang/srt/models/dspark.py` 中加载 `d2t` 和独立的 `lm_head.weight`。
- 按 checkpoint 的实际词表大小建立 Markov 输入 embedding，保留缩减词表的输出维度。
- 使用 `target_id = draft_id + d2t[draft_id]` 恢复目标 token ID。
- 将 draft logits 和 Markov bias 映射至目标词表；候选词表外的 base logits 设置为 `-inf`，保持零采样概率。
- 检查映射类型、重复或越界 ID、权重形状、缺失 head 和目标词表一致性。
- 无 `d2t` 的 full-vocabulary draft 保留原有 target-sharing 路径。

**结果：** 旧 draft 可以完成 TP4 启动和全量精度评测。适配阶段的 7 项回归测试通过，覆盖权重加载、数值计算、全词表路径、非法输入及 DCU Graph 捕获/回放。

### 2. `8bb8ae250c` — 兼容 compact draft 配置与辅助层字段

Commit：`8bb8ae250c504d7f0329200db98919de8aea1012`

`Support compact DSPARK draft configurations and auxiliary layer aliases`

**问题：** 新 draft 将网络参数放在 `transformer_layer_config` 中，顶层缺少普通 Hugging Face 配置所需的 `model_type` 等字段，直接加载会报 `Unrecognized model ... Should have a model_type key`。其辅助层选择字段为 `aux_hidden_state_layer_ids`，需要被正确识别并保留顺序。

**修改：**

- 新增 `python/sglang/srt/configs/dspark.py`，将 compact DSPARK 配置转换为可加载的平面 HF 网络配置。
- 在 `python/sglang/srt/utils/hf_transformers/config.py` 的统一加载入口应用转换；不改写 checkpoint 文件，旧平面配置保留原加载流程。
- 保留网络层类型、RoPE 参数和 draft 元数据；校验必要字段及内外层冲突，避免外层空值覆盖有效网络参数。
- 在 `python/sglang/srt/speculative/dflash_utils.py` 中兼容 `aux_hidden_state_layer_ids`，保留辅助层顺序；与 `target_layer_ids` 冲突时明确报错。

**结果：** 新旧配置等 9 项测试通过。新 draft 的 `fc.weight=[2048,16384]` 与 8 个辅助层一致；checkpoint 不含 `d2t`、embedding 和 LM head，符合其 target-sharing 导出方式。完整服务启动还需下一个提交解决 FA3 MTP 内核限制。

### 3. `48fcb0299b` — 修复 HCU FA3 MTP 头数限制

Commit：`48fcb0299bbeb744dd7e07e436aa7672878edd4d`

`Split HCU causal MTP queries exceeding paged attention head limits`

**问题：** 支持draft model 的 `block_size=15`，target verify 每请求处理 16 个 token。TP2 启动时权重与 FP8 KV Cache 已加载成功，但 CUDA Graph 捕获触发 HCU paged attention 的 `num_heads <= num_kv_heads * 64` 断言。该内核将 MTP 查询 token 折叠到头维度，限制实际涉及 `query_heads × query_length`。

**修改：**

- 在 `python/sglang/srt/layers/attention/flashattention_interface.py` 中，对超过限制的均匀因果查询按 token 分块调用同一 FA3 接口。
- 根据每块在原序列中的位置调整可见 KV 长度，保持右下对齐的因果语义，并拼接恢复原输出顺序。
- 分块大小由 Q/KV 头数和内核限制计算，不绑定具体 draft 权重。
- 保留 FP8 KV Cache 和 CUDA Graph，不通过缩短模型 block size 或关闭 graph 绕过问题。

新增路径限定于 HCU、legacy BHSD、page size 64、无滑窗、查询长度 2～16 的均匀因果 MTP，且满足相应 head dimension 与单 token 头数限制。其他形状沿用既有路径。

**结果：** 9、15、16 token 的 FP8 KV attention 对照 PyTorch 参考计算及 CUDA Graph 回放全部通过，最大绝对误差为 `0.00390625`。新 draft 的 TP2 服务成功启动，并完成单请求、流式、32 并发和全量精度评测。

### 公共测试配置

- 主模型与 draft 均使用 FA3、FP8 E5M2 KV Cache；`SGLANG_ROCM_USE_AITER_MOE=1`。
- CUDA Graph 开启，target/draft verify 捕获成功。
- HumanEval：164 条，指标为 Pass@1；GSM8K：1319 条，指标为 Accuracy。
- `enable_thinking=false`、`temperature=0`、`top_p=1`、`seed=42`、`max_tokens=13312`。
- Eval batch size：32；流式 OpenAI-compatible API；HumanEval `review_timeout=30`。

### draft：Qwen3.6-35B-A3B-speculator.dspark

配置：TP4，DCU 4/5/6/7，block size 8，target verify 9 token；验证代码对应 `2c6b570779`。

- **HumanEval Pass@1：95.12%（156/164）**；8 题未通过，执行错误 0。
- **GSM8K Accuracy：96.89%（1278/1319）**；41 题未通过，执行错误 0。

## Speed Tests and Profiling

本次完成了服务运行与 CUDA Graph 验证，**尚未完成相同 TP、相同输入下的无投机性能基线对比，也未采集 profiler trace，因此不声明加速收益**。

已验证：

- draft model 在 TP4下成功捕获 target verify 和 draft verify CUDA Graph。
- 实际解码日志确认 CUDA Graph 正常使用。
- FA3 分块修复继续调用原 FA3 内核，但额外分块、复制和拼接的性能开销尚未独立量化。

待补充的性能验证：固定 TP、输入/输出长度和并发，比较无投机基线与 DSPARK 的吞吐、TTFT、TPOT、接受长度，并量化 FA3 分块路径的开销。

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

当前状态说明：

- 已执行 `git diff --check` 和配套 shell 脚本语法检查，未执行完整 pre-commit 检查。
- 已运行上述回归、数值与 API 检查。
- 全量精度结果已提供；独立速度 benchmark 和 profiling 尚未完成，因此合并项保持未勾选。
- 尚未完成依据贡献指南的完整代码风格核验，Checklist 不标记为已完成。

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

以上为后续审阅流程；本说明不代表已取得审批、通过 CI 或完成合并。

---
### CI States

Latest PR Test (Base): :x: **Missing `run-ci` label** -- add it to run CI tests.
Latest PR Test (Extra): :x: **Blocked** -- `run-ci` is required first.

### CODE REVIEW

**Comment by chenyuege** [comment]
启动脚本的部分配置如下：
 --kv-cache-dtype fp8_e5m2 --trust-remote-code
    --speculative-algorithm DSPARK
    --speculative-draft-model-path "$DRAFT_MODEL_PATH"
    --speculative-num-steps 1 --speculative-eagle-topk 1
    --speculative-dspark-block-size 8
    --speculative-draft-attention-backend fa3
    --speculative-draft-kv-cache-dtype fp8_e5m2

**Review by lizhigong1988 — APPROVED** [review]
